src/executor/current_thread/mod.rs

```diff
@@ -0,0 +1,701 @@
+}
+
+/// Executes a future on the current thread.
+///
+/// The provided future must complete or be canceled before `run` will return.
+///
+/// # Panics
+///
+/// This function can only be invoked from the context of a `run` call; any
+/// other use will result in a panic.
+pub fn spawn<F>(future: F)
+where F: Future<Item = (), Error = ()> + 'static
+{
+    TaskExecutor::current()
+        .spawn_local(Box::new(future))
+        .unwrap();
+}
+
+// ===== impl CurrentThread =====
+
+impl CurrentThread<ParkThread> {
+    /// Create a new instance of `CurrentThread`.
+    pub fn new() -> Self {
+        CurrentThread::new_with_park(ParkThread::new())
+    }
+}
+
+impl<P: Park> CurrentThread<P> {
+    /// Create a new instance of `CurrentThread` backed by the given park
+    /// handle.
+    pub fn new_with_park(park: P) -> Self {
+        let unpark = park.unpark();
+
+        CurrentThread {
+            scheduler: Scheduler::new(unpark),
+            num_futures: 0,
+            park,
+        }
+    }
+
+    /// Returns `true` if the executor is currently idle.
+    ///
+    /// An idle executor is defined by not currently having any spawned tasks.
+    pub fn is_idle(&self) -> bool {
+        self.num_futures == 0
+    }
+
+    /// Spawn the future on the executor.
+    ///
+    /// This internally queues the future to be executed once `run` is called.
+    pub fn spawn<F>(&mut self, future: F) -> &mut Self
+    where F: Future<Item = (), Error = ()> + 'static,
+    {
+        self.borrow().spawn_local(Box::new(future));
+        self
+    }
+
+    /// Synchronously waits for the provided `future` to complete.
+    ///
+    /// This function can be used to synchronously block the current thread
+    /// until the provided `future` has resolved either successfully or with an
+    /// error. The result of the future is then returned from this function
+    /// call.
+    ///
+    /// Note that this function will **also** execute any spawned futures on the
+    /// current thread, but will **not** block until these other spawned futures
+    /// have completed.
+    ///
+    /// The caller is responsible for ensuring that other spawned futures
+    /// complete execution.
+    pub fn block_on<F>(&mut self, future: F)
+        -> Result<F::Item, BlockError<F::Error>>
+    where F: Future
+    {
+        let mut enter = tokio_executor::enter().unwrap();
+        self.enter(&mut enter).block_on(future)
+    }
+
+    /// Run the executor to completion, blocking the thread until **all**
+    /// spawned futures have completed.
+    pub fn run(&mut self) -> Result<(), RunError> {
+        let mut enter = tokio_executor::enter().unwrap();
+        self.enter(&mut enter).run()
+    }
+
+    /// Run the executor to completion, blocking the thread until all
+    /// spawned futures have completed **or** `duration` time has elapsed.
+    pub fn run_timeout(&mut self, duration: Duration)
```

Comment:
General question: why does Rust prefer timeouts vs deadlines? Timeouts don't compose well:

```rust
// I want to finish something in 10 secs, with timeouts.
let mut start = Instant::now();
let elapsed = || {
  let now = Instant::now();
  let res = start - res;
  start = now;
  res
};
// Not even entirely correct because elapsed() can be greater than 10 secs and subtraction will panic.
let c = socket.open_with_timeout(Duration::from_secs(10))?;
let data = socket.read_with_timeout(Duration::from_secs(10) - elapsed())?;
let processed = process(data)?;
socket.write_with_timeout(processed, Duration::from_secs(10) - elapsed())?;
```

```rust
// I want to finish something in 10 secs, with deadline.
let deadline = Instant::now() + Duration::from_secs(10);
let c = socket.open_with_deadline(deadlline)?;
let data = socket.read_with_deadline(deadlline)?;
let processed = process(data)?;
socket.write_with_deadline(processed, deadlline)?;
```

Reply:
IIRC, Rust itself picked timeouts as that is what gets passed to the syscalls.

One can add a deadline shim on top of a timeout based API, but if the APIs themselves are deadline based, you would have to make additional syscalls within each one to get `Instant::now()`.

I personally have never hit the composability issue as working w/ futures it is a non-issue (you define the computation w/o any timeout or deadline and set a timeout on the entire computation). However, if it did become an issue, I'd probably add a deadline shim on top :)

Reply:
I think adding a timeout shim on top of deadline API is easier (no need to check for negative timeouts). Also deadline is the right thing for the implementation anyway.

I don't think there is additional `Instant::now()` call unless the timeout based code has bugs (see below).

The composability issue is a real problem. Also timeouts usually lead to bugs whenever there are retries or partial reads. Go switched from SetTimeout to SetDeadline [before 1.0](https://github.com/golang/go/commit/b71883e9b0eff7e89081d20204bf33f369cdf735) mainly for the latter reason.

FWIW I just check stdlib and it suffers from this as well. If we set a timeout on a socket and then use [`read_to_end`](https://doc.rust-lang.org/stable/src/std/io/mod.rs.html#366-395) or [`read_exact`](https://doc.rust-lang.org/stable/src/std/io/mod.rs.html#701-716) the timeout won't be obeyed if the loops inside said functions execute more than once. Even if we try to describe the current semantics they are super weird: the effective timeout is K * the set timeout where K is the number of successful partial reads of `read_exact` or `read_to_end` respectively.

Reply:
@alkis For me, it comes down to this. How do you change [this](https://github.com/tokio-rs/tokio/blob/tokio-runtime/tokio-executor/src/park.rs#L231-L271) function to take a deadline instead of a timeout without increasing the number of syscalls?

Reply:
@carllerche the function you linked will gain one `Instant::now` and [this](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/condvar.rs#L87) function will lose [this](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/condvar.rs#L91) syscall.

Reply:
Looks like you are correct re that one, but other platforms take [timeouts](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686304(v=vs.85).aspx). Also `epoll_wait` takes a timeout 🤷‍♀️ .

Either way, Rust picked its idiom and I prefer sticking w/ the idioms.

Reply:
Rust can support both timeout and deadline and each platform's implementation can pick whichever one works best.

In either case, `read_to_end` and `read_exact` need to be fixed.

Reply:
Here is an informative article about this issue (Python-oriented but relevant for any language I think): https://vorpus.org/blog/timeouts-and-cancellation-for-humans/

Regarding `read_to_end` and friends, there are actually two behaviors you might want:

- A timeout on inactivity - "keep going as long as you received at least one byte in the last 5 seconds" - which appears to be what is currently (inadvertently?) implemented.

- A deadline - "keep going until 15:30:53.35353".

A deadline is what most situations want, though.

Reply:
Anyone wants to take this up on an RFC? At this point there are only 20 functions that take timeout in the stdlib so it is not an insurmountable amount of work.

Reply:
Timeouts work better on systems with non-monotic clocks, or if you have got multiple clocks on-board. if clocks are set back or forward in time, deadlines are causing big troubles.

Reply:
@frehberg Can elaborate why deadlines are causing troubles on non-monotonic clocks? Also what kind of systems these apply to?

Reply:
@alkis for example  in automotive, having multiple internal clocks, such as  the dashboard-clock,  power-control-unit-clock,  system-clock, and steady-clock. The latter being a monotonic clock; the time points of this clock cannot decrease as physical time moves forward. In contrast a  system clock may be  adjusted by user or via GPS. If you define a deadline-timer relative to the system clock, and in the meanwhile the clock is adjusted to a timepoint in the past, the timer will fire after a much longer interval than intended; the corresponding  functionality of the app may seem frozen and  malfunctioning.

Reply:
That's sounds like a bug. There is access to monotonic clock yet one is defined on a non monotonic clock.

I am looking for an example where the system does not have a monotonic clock and the program specifies a timeout and that is handled correctly.

Reply:
FYI I wrote [this](https://www.reddit.com/r/rust/comments/8b5krv/the_case_for_deadlines/) on the topic.

Reply:
@alkis you probably saw, but tokio-timer is now using deadlines.

Reply:
No, I didn't. Thanks for the pointer @carllerche!